c10/core/TensorImpl.cpp

```diff
@@ -122,12 +122,34 @@ TensorImpl::TensorImpl(
   }
 }
 
+// This is currently used to construct view tensors for nested_tensor impls
 TensorImpl::TensorImpl(
+    ImplType type,
     DispatchKeySet key_set,
     const caffe2::TypeMeta data_type,
     c10::optional<c10::Device> device_opt)
     // NOLINTNEXTLINE(performance-move-const-arg)
-    : TensorImpl({}, key_set, data_type, std::move(device_opt)) {}
```

Comment:
couldn't reuse the constructor on line: 105 because this takes the device type from storage.. sigh.. so created a new constructor

Reply:
You should use this overload directly

```
TensorImpl::TensorImpl(
    Storage&& storage,
    DispatchKeySet key_set,
    const caffe2::TypeMeta data_type,
    c10::optional<c10::Device> device_opt)
```

If you want to be a good citizen, you should refactor 

```
TensorImpl::TensorImpl(
    ImplType type,
    Storage&& storage,
    DispatchKeySet key_set,
    const caffe2::TypeMeta data_type)
```

to call into the above constructor too.

Reply:
With the first overload you mentioned though can't I not set the ImplType::VIEW enum though? I think the problem is that the body of the first construct does alot of extra work like removing keys when in inference mode and the second constructor is used to avoid all the extra work. Maybe this is a c++ thing I am confused on

Reply:
There is an efficiency thing going on here, but it's not relevant to your PR.

This constructor, as ImplType enum suggests, is specifically for constructing dense-to-dense views.

```
  // Note [Enum ImplType]
  // This enum is temporary. In the followup refactor we should
  // think about how to specialize TensorImpl creation for view
  // tensors. Currently we only special case its key_set_ but
  // there's also potential to share version_counter_ directly
  // without creating first and then override in as_view.
  enum ImplType { VIEW };
```

It came out of inference mode related changes, see #54403. The primary performance implication is not about skipping dispatch key manipulation (cheap) but avoiding dynamic allocations for view counters. I suppose I would not suggest you attempt this refactor, though there is a refactor opportunity as the comment suggests.

In any case, you shouldn't be touching any of this machinery in your PR. The view constructor was specifically designed storage-based tensors (which nested tensor is not), and if you want to generalize it for nested tensor there needs to be a design. But I don't think any of this matters. Just use the stock non-view constructor.